### ISSUE TITLE
dyn-star: "needs to be a pointer-sized type" but the type is pointer-sized

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#107696]
Consider this code:
```rust
#![feature(dyn_star)]
#![allow(incomplete_features)]

use std::fmt::Debug;

fn main() {
    let _dyn_i: dyn* Debug = [0u8; std::mem::size_of::<usize>()];
}
```
This raises an error:
```
error[E0277]: `[u8; 8]` needs to be a pointer-sized type
```
However, it *is* a pointer-sized type. Looks like somehow that's still not good enough?

On a related note, I am surprised that this `PointerSize` marker trait (according to its documentation) cares only about size and not alignment. I think things would go very wrong if we had a type that had ptr size but larger alignment than a ptr, and then turned that into a `dyn*`...

**Comment by RalfJung** [comment]
I think I found the code implementing this:

https://github.com/rust-lang/rust/blob/3de7d7fb22a579a3d59ddb1c959d1b3da224aafa/compiler/rustc_trait_selection/src/solve/trait_goals.rs#L134-L155

So actually, this trait is about the type having both the size *and alignment* of a pointer type. That resolves my alignment-related concerns, but means that the doc comment and error message associated with `PointerSize` (and its name!) are missing part of the story.
